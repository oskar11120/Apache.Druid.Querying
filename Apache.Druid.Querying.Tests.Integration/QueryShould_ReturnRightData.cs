using FluentAssertions;
using Snapshooter;
using Snapshooter.NUnit;
using System.Globalization;
using System.Runtime.CompilerServices;
using static Apache.Druid.Querying.Tests.Integration.ServiceProvider;

namespace Apache.Druid.Querying.Tests.Integration;

internal static class TestData
{
    public static readonly DateTimeOffset T0 = DateTimeOffset.Parse("2016-06-27T00:00:00.000Z", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal);
    public static readonly Interval Interval = new(T0, T0.AddDays(1));

    public static TQuery DefaultInterval<TQuery>(this TQuery query)
        where TQuery : IQueryWith.Intervals
        => query
            .Interval(Interval);
}

internal class QueryShould_ReturnRightData
{
    private static async Task VerifyMatch<TSource, TResult, TContext>(
        DataSource<TSource> dataSource,
        IQueryWithSource<TSource>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
    {
        var spanshotName = Snapshot.FullName(new SnapshotNameExtension(snapshotNameSuffix));
        var json = dataSource.MapQueryToJson(query).ToString();
        await TestContext.Out.WriteLineAsync(json);
        void Match(List<TResult> results)
            => Snapshot.Match(results, spanshotName, options => options.IgnoreField("[:].SegmentId"));
        var results = await dataSource
            .ExecuteQuery(query)
            .ToListAsync();
        Match(results);

        var withNoTruncatedResultHandling = await dataSource
            .ExecuteQuery(query, onTruncatedResultsQueryRemaining: false)
            .ToListAsync();
        Match(withNoTruncatedResultHandling);
    }

    private static Task VerifyMatch<TResult, TContext>(
        IQueryWithSource<Edit>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
        => VerifyMatch(Wikipedia.Edits, query, snapshotNameSuffix);

    [Test]
    public async Task Wrapped()
    {
        var first = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>()
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .DefaultInterval()
            .Granularity(Granularity.FifteenMinutes)
            .Context(new() { SkipEmptyBuckets = true });
        var wrapped = Wikipedia
            .Edits
            .ToQueryDataSource(first);
        var query = new Query<WithTimestamp<TimeSeriesAggregations>>
            .Scan()
            .Interval(TestData.Interval)
            .Filter(type => type.Not(type.Selector(data => data.Value.Count, 6)))
            .Limit(1000);
        await VerifyMatch(wrapped, query, string.Empty);
    }

    public enum JoinTestCase
    {
        LatestEditTimestampsPerCountry,
        LatestEditsPerCountry,
        Both
    }
    private record Country(string Code, string FullName);
    [TestCase(JoinTestCase.LatestEditTimestampsPerCountry)]
    [TestCase(JoinTestCase.LatestEditsPerCountry)]
    [TestCase(JoinTestCase.Both)]
    public async Task Join(JoinTestCase @case)
    {
        var latestEditTimestampsPerCountry_query = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<DateTimeOffset>()
            .DefaultInterval()
            .Granularity(Granularity.Day)
            .Dimensions(type => type.Default(data => data.CountryName))
            .Aggregations(type => type.Last(data => data.Timestamp));
        var latestEditTimestampsPerCountry_dataSource = Wikipedia.Edits.ToQueryDataSource(latestEditTimestampsPerCountry_query);
        var lastestEditsPerCountry_dataSource = Wikipedia.Edits.InnerJoin(
            latestEditTimestampsPerCountry_dataSource,
            data => $"{data.Left.CountryName} == {data.Right.Value.Dimensions} && {data.Left.Timestamp} == {data.Right.Value.Aggregations}");
        var edits_query = new Query<InnerJoinData<Edit, WithTimestamp<Dimensions_Aggregations<string, DateTimeOffset>>>>
            .Scan()
            .DefaultInterval();
        await (@case switch
        {
            JoinTestCase.LatestEditTimestampsPerCountry => VerifyMatch(Wikipedia.Edits, latestEditTimestampsPerCountry_query, @case.ToString()),
            JoinTestCase.LatestEditsPerCountry => VerifyMatch(lastestEditsPerCountry_dataSource, edits_query, @case.ToString()),
            JoinTestCase.Both => VerifyBoth(),
            _ => throw new NotSupportedException()
        });

        async Task VerifyBoth()
        {
            var fromEdits = await lastestEditsPerCountry_dataSource
                .ExecuteQuery(edits_query)
                .Select(result => result.Value.Right.Value)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            var fromTimestamps = await Wikipedia.Edits
                .ExecuteQuery(latestEditTimestampsPerCountry_query)
                .Select(result => result.Value)
                .Where(result => result.Dimensions is not null)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            fromEdits.Should().BeEquivalentTo(fromTimestamps);
        }
    }

    [Test]
    public async Task Union()
    {
        var union = Wikipedia
            .Edits
            .Union(Wikipedia
                .Edits);
        var query = new Query<Union<Edit, Edit>>
            .Scan()
            .Interval(TestData.Interval)
            .Limit(10);
        await VerifyMatch(union, query, string.Empty);
    }

    private sealed record InlineData(string Word, [property: DataSourceColumn("Integer")] int Number);
    [Test]
    public async Task Inline()
    {
        var query = new Query<InlineData>
            .Scan()
            .Interval(new(default, DateTimeOffset.MaxValue))
            .Filter(type => type.In(
                data => data.Word,
                new[] { "zero" }));
        var inline = Wikipedia
            .Inline(new InlineData[]
            {
                new("zero", 1),
                new("one", 2)
            });
        var result = await inline
            .ExecuteQuery(query)
            .ToListAsync();
        result
            .Single()
            .Value
            .Should()
            .Be(new InlineData("zero", 1));
    }

    [Test]
    public async Task Scan()
    {
        var first3Us = new Query<Edit>
            .Scan()
            .DefaultInterval()
            .Filter(type => type.Selector(
                edit => edit.CountryIsoCode,
                "US"))
            .Limit(3);
        await VerifyMatch(first3Us);
    }

    [Test]
    public async Task GroupBy()
    {
        var countsPerCountry = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .DefaultInterval()
            .Dimensions(type => type.Default(edit => edit.CountryName))
            .Aggregations(type => type.Count())
            .Granularity(Granularity.EightHours);

        var countsPerCountryOrderedByCount = countsPerCountry
            .LimitSpec(limit: 100, offset: 1, columns => columns.OrderBy(data => data.Aggregations, OrderDirection.Descending, SortingOrder.Numeric));
        await VerifyMatch(countsPerCountryOrderedByCount);

        var countsPerValidCountry = countsPerCountry
            .Filter(type => type.Not(type.Selector(edit => edit.CountryName, string.Empty)));
        await VerifyMatch(countsPerValidCountry);
    }

    [Test]
    public async Task TopN()
    {
        var countsPerPage = new Query<Edit>
            .TopN<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .Dimension(type => type.Default(edit => edit.Page))
            .Threshold(10)
            .Aggregations(type => type.Count())
            .Metric(type => type.Numeric(data => data.Aggregations))
            .DefaultInterval()
            .Granularity(Granularity.Hour);
        await VerifyMatch(countsPerPage);

        var usAnonymousCountsPerPage = countsPerPage
            .Filter(type => type.And(
                type.Selector(edit => edit.IsAnonymous, true),
                type.Selector(edit => edit.CountryIsoCode, "US")));
        await VerifyMatch(usAnonymousCountsPerPage);
    }

    private record TimeSeriesAggregations(int Count, int TotalAdded);
    private record TimeSeriesPostAggrgations(double AverageAdded);
    [Test]
    public async Task TimeSeries()
    {
        var lineStatisticsPerHour = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>
            .WithPostAggregations<TimeSeriesPostAggrgations>()
            .Order(OrderDirection.Descending)
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .PostAggregations(type => new(type.Arithmetic(
                ArithmeticFunction.Divide,
                type.FieldAccess(aggregations => aggregations.TotalAdded),
                type.FieldAccess(aggregations => aggregations.Count))))
            .Filter(type => type.Selector(edit => edit.CountryIsoCode, "US"))
            .DefaultInterval()
            .Granularity(Granularity.Hour)
            .Context(new QueryContext.TimeSeries() { SkipEmptyBuckets = true });
        await VerifyMatch(lineStatisticsPerHour);

        var inEuWarsawTime = lineStatisticsPerHour
            .Granularity(Granularity.Hour, "Europe/Warsaw");
        await VerifyMatch(inEuWarsawTime);
    }


    private sealed record GroupByBooleans(bool Robot, bool New);
    [TestCase(0)]
    [TestCase(1)]
    public async Task WithDataPropertiesAccessedByInterface(int i)
    {
        async Task Test<TEdit>(DataSource<TEdit> dataSource) where TEdit : IEditBooleans
        {
            var one = new Query<TEdit>
                .Scan()
                .DefaultInterval()
                .Filter(type => type.And(
                    type.Selector(data => data.IsNew, true),
                    type.Selector(data => data.Robot, true)))
                .Limit(100);
            if (i is 0)
                await VerifyMatch(dataSource, one);

            var two = new Query<TEdit>
                .GroupBy<GroupByBooleans>
                .WithNoVirtualColumns
                .WithAggregations<int>()
                .DefaultInterval()
                .Granularity(Granularity.Hour)
                .Dimensions(type => new(
                    type.Default(data => data.Robot),
                    type.Default(data => data.IsNew)))
                .Aggregations(type => type.Count());
            if (i is 1)
                await VerifyMatch(dataSource, two);
        }

        await Test(Wikipedia.Edits);
    }

    interface IIotMeasurement
    {
        Guid ObjectId { get; }
        double Value { get; }
    }
}
