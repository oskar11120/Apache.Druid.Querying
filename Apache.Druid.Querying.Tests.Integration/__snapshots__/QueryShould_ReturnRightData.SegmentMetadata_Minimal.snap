[
  {
    "Id": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-03-06T22:10:26.236Z",
    "Columns": {
      "__time": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "isRobot": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "false",
        "MaxValue": "true",
        "Cardinality": 2,
        "ErrorMessage": null
      },
      "channel": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "#ar.wikipedia",
        "MaxValue": "#zh.wikipedia",
        "Cardinality": 51,
        "ErrorMessage": null
      },
      "flags": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "",
        "MaxValue": "NB",
        "Cardinality": 10,
        "ErrorMessage": null
      },
      "isUnpatrolled": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "false",
        "MaxValue": "true",
        "Cardinality": 2,
        "ErrorMessage": null
      },
      "page": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "'t Suydevelt",
        "MaxValue": "황소자리 세타",
        "Cardinality": 22383,
        "ErrorMessage": null
      },
      "diffUrl": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "https://ar.wikipedia.org/w/index.php?diff=20310994&oldid=20310977",
        "MaxValue": "https://zh.wikipedia.org/w/index.php?title=Topic:T6oo8x0xa2kghw9d&action=history",
        "Cardinality": 24433,
        "ErrorMessage": null
      },
      "added": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "comment": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "!vote Keep",
        "MaxValue": "（简体中文）. 引文格式1维护：未识别语文类型 (link)其他特色條目：綏遠抗戰 － 巴伐利亚级战列舰 － 第一代纳尔逊子爵霍雷肖·纳尔逊 特色列表：印度尼西亚总理列表 － 苏联最高领导人列",
        "Cardinality": 13865,
        "ErrorMessage": null
      },
      "commentLength": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "isNew": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "false",
        "MaxValue": "true",
        "Cardinality": 2,
        "ErrorMessage": null
      },
      "isMinor": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "false",
        "MaxValue": "true",
        "Cardinality": 2,
        "ErrorMessage": null
      },
      "delta": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "isAnonymous": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "false",
        "MaxValue": "true",
        "Cardinality": 2,
        "ErrorMessage": null
      },
      "user": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "*feridiák",
        "MaxValue": "ＫｏＺ",
        "Cardinality": 7862,
        "ErrorMessage": null
      },
      "deltaBucket": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "deleted": {
        "TypeSignature": "LONG",
        "Type": "LONG",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": null,
        "MaxValue": null,
        "Cardinality": null,
        "ErrorMessage": null
      },
      "namespace": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": false,
        "Size": 0,
        "MinValue": "Aide",
        "MaxValue": "틀토론",
        "Cardinality": 235,
        "ErrorMessage": null
      },
      "cityName": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "Zurich",
        "Cardinality": 836,
        "ErrorMessage": null
      },
      "countryName": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "Vietnam",
        "Cardinality": 106,
        "ErrorMessage": null
      },
      "regionIsoCode": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "ZP",
        "Cardinality": 354,
        "ErrorMessage": null
      },
      "metroCode": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "866",
        "Cardinality": 92,
        "ErrorMessage": null
      },
      "countryIsoCode": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "ZA",
        "Cardinality": 106,
        "ErrorMessage": null
      },
      "regionName": {
        "TypeSignature": "STRING",
        "Type": "STRING",
        "HasMultipleValues": false,
        "HasNulls": true,
        "Size": 0,
        "MinValue": null,
        "MaxValue": "Świętokrzyskie",
        "Cardinality": 470,
        "ErrorMessage": null
      }
    },
    "NumRows": 24433,
    "Size": 0,
    "Intervals": [
      {
        "From": "2016-06-27T00:00:00+00:00",
        "To": "2016-06-28T00:00:00+00:00"
      }
    ],
    "TimestampSpec": null,
    "Aggregators": null,
    "QueryGranularity": null,
    "Rollup": null
  }
]
